### ISSUE TITLE
Requires libsgx_urts.so to run in SGX hardware mode

### ISSUE THREAD

**Issue by nhoriguchi** [hyperledger-labs/minbft#93]
Current installation procedure seems to cover only cases of running in SGX simulation mode. I saw the following error in running `keytool generate` with `SGX_MODE=HW`:

> bin/keytool: error while loading shared libraries: libsgx_urts.so: cannot open shared object file: No such file or directory

`libsgx_urts_sim.so` for simulation mode is installed by SGX SDK installer `sgx_linux_x64_sdk_2.3.101.46683.bin`, and `libsgx_urts.so` for hardware mode is delivered by `libsgx-enclave-common_2.3.101.46683-1_amd64.deb`. So we might be better to add some instruction for users who want to run MinBFT in hardware mode.

**Comment by ynamiki** [comment]
Right, we need to install a driver and a platform software (PSW) in addition to the SDK to run the command in the hardware mode. We can update the README to ask a user to install these packages based on [the Intel's Installation Guide](https://download.01.org/intel-sgx/linux-2.3.1/docs/Intel_SGX_Installation_Guide_Linux_2.3.1_Open_Source.pdf).

@Naoya-Horiguchi Are you creating a patch for this? Shall I do this?

**Comment by nhoriguchi** [comment]
@ynamiki, I'll work on this.

**Comment by sergefdrv** [comment]
Please check if we actually need to install PSW. MinBFT doesn't seem to use any of PSW features. If I remember correctly, it was enough to install SGX driver prior to SDK to make hardware libraries available.

**Comment by nhoriguchi** [comment]
@sergefdrv, thanks for the comment. Actually `libsgx_urts.so` is included in Intel SGX SDK, so installing PSW is not mandatory. It seems that the problem is that `libsgx_urts.so` and `libsgx_urts_sim.so` are located under different directories, and only `libsgx_urts_sim.so` is available under the library path `/opt/intel/sgxsdk/sdk_libs` (specified in `/etc/ld.so.conf.d/sgx-sdk.conf`):

> root@sgx3:~/tmp# find /opt/intel | grep urts
> /opt/intel/sgxsdk/include/sgx_urts.h
> /opt/intel/sgxsdk/lib64/libsgx_urts_sim.so
> /opt/intel/sgxsdk/lib64/libsgx_urts.so
> /opt/intel/sgxsdk/sdk_libs/libsgx_urts_sim.so
> /opt/intel/sgxsdk/pkgconfig/libsgx_urts_sim.pc
> /opt/intel/sgxsdk/pkgconfig/libsgx_urts.pc

So adding another path `/opt/intel/sgxsdk/lib64` to `/etc/ld.so.conf.d/sgx-sdk.conf` might improve the situation?

**Comment by sergefdrv** [comment]
@Naoya-Horiguchi thanks for investigating that. I think it would be enough to simply change `/opt/intel/sgxsdk/sdk_libs` to `/opt/intel/sgxsdk/lib64` in `/etc/ld.so.conf.d/sgx-sdk.conf` in order to support both simulation- and hardware-based USIG.

**Comment by nhoriguchi** [comment]
When I tried to include `/opt/intel/sgxsdk/lib64` into library path, I encountered an interesting error on running `usig_test`:
~~~
RUN    test/usig_test                                                                    
Please use the correct uRTS library from PSW package.                                    
usig_test: test/usig_test.c:38: test_init_destroy: Assertion `usig_init(enclave_file, &ei, NULL, 0) == SGX_SUCCESS' failed.                                                       
~~~
This error message is self-descriptive and also mentioned [here](https://github.com/intel/linux-sgx/issues/47). We have to use the reliable version distributed with PSW library and can't use one in SGX SDK. So my current approach doesn't work :worried:. 

Another point is that I think we don't need to create `/etc/ld.so.conf.d/sgx-sdk.conf` because that doesn't become effective without running `ldconfig` command to update `/etc/ld.so.cache`. For testing purpose in simulation mode, `LD_LIBRARY_PATH` is updated in `$SGX_SDK/environment` and that's enough.

If these points sound OK, I'll update related part in README.

**Comment by sergefdrv** [comment]
@Naoya-Horiguchi Thanks for the analysis. Indeed, we seem to have to install PSW, then. So we need to mention this in Readme file.

> `LD_LIBRARY_PATH` is updated in `$SGX_SDK/environment` and that's enough

It might  be a bit inconvenient to invoke `$SGX_SDK/environment`each time. So it might be useful to keep the optional advise to create `/etc/ld.so.conf.d/sgx-sdk.conf`. Let's just mention that `sudo ldconfig` has to be invoked once after adding the file.

**Comment by nhoriguchi** [comment]
> So it might be useful to keep the optional advise to create /etc/ld.so.conf.d/sgx-sdk.conf. Let's just mention that sudo ldconfig has to be invoked once after adding the file.

@sergefdrv I'll update #95 with this direction. Thanks for sharing ideas.

**Comment by nhoriguchi** [comment]
closes by merging #96.
